### PR TITLE
cli,server: distinguish graceful vs non-graceful shutdown requests

### LINKED ISSUE #107856: ccl/serverccl: TestServiceShutdownUsesGracefulDrain failed
ccl/serverccl.TestServiceShutdownUsesGracefulDrain [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11115347?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11115347?buildTab=artifacts#/) on master @ [f295bd861a3a427652b19c2254d2401ebb4a3c8e](https://github.com/cockroachdb/cockroach/commits/f295bd861a3a427652b19c2254d2401ebb4a3c8e):

```
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !goroutine 27787625 [sync.Cond.Wait]:
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !sync.runtime_notifyListWait(0xc013189690, 0x20e)
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	GOROOT/src/runtime/sema.go:517 +0x14c
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !sync.(*Cond).Wait(0x58c8f60?)
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	GOROOT/src/sync/cond.go:70 +0x8c
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker(0xc0029f58f0, {0x79c1540, 0xc021bb24b0}, {0x79a4c20, 0xc003768a80}, 0xc003870a80)
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:378 +0x318
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2({0x79c1540?, 0xc021bb24b0?})
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:321 +0x45
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484 +0x146
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:475 +0x43b
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !goroutine 27787367 [sync.Cond.Wait]:
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !sync.runtime_notifyListWait(0xc013189610, 0x356)
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	GOROOT/src/runtime/sema.go:517 +0x14c
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !sync.(*Cond).Wait(0x58c8f60?)
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	GOROOT/src/sync/cond.go:70 +0x8c
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker(0xc0029f56c0, {0x79c1540, 0xc02de2fad0}, {0x79a4c20, 0xc003768a80}, 0xc003870a80)
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:378 +0x318
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2({0x79c1540?, 0xc02de2fad0?})
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:321 +0x45
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484 +0x146
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:475 +0x43b
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !For more context, check log files in: /artifacts/tmp/_tmp/c101b7a464a1afc1f5af0cd85792187e/logTestServiceShutdownUsesGracefulDrain2336992875
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !****************************************************************************
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !This node experienced a fatal error (printed above), and as a result the
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !process is terminating.
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !Fatal errors can occur due to faulty hardware (disks, memory, clocks) or a
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !problem in CockroachDB. With your help, the support team at Cockroach Labs
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !will try to determine the root cause, recommend next steps, and we can
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !improve CockroachDB based on your report.
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !Please submit a crash report by following the instructions here:
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !    https://github.com/cockroachdb/cockroach/issues/new/choose
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !If you would rather not post publicly, please contact us directly at:
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !    support@cockroachlabs.com
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !The Cockroach Labs team appreciates your feedback.
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/multi-tenant @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestServiceShutdownUsesGracefulDrain.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-30222
Epic: CRDB-26691

### LINKED ISSUE #108611: cli/start: improper graceful drain upon stop trigger shutdown requests
**Describe the problem**

The shutdown logic in cli/start tries graceful drains when receiving shutdown requests. This is incorrect; currently all shutdown reasons require non-graceful shutdowns.

**Expected behavior**

For all the current reasons the code in cli/start should use a non-graceful shutdown.

Epic: CRDB-28893

Jira issue: CRDB-30548

### PR DESCRIPTION
**Pull request by knz**
Needed to fix #107856.

TLDR: this patch fixes a bug whereby the top-level shutdown logic in `cli/start.go` mistakenly attempted a graceful drain when that, in fact, was impossible. This would show up as a storm of errors in logs and, sometimes, block server shutdown until a non-graceful shutdown was initiated by other means.

A more detailed explanation follows.

*For context*, we have a notification mechanism (called "stop trigger") through which internal components in a server can request a shutdown of the overall server. As of this writing, this mechanism is used for example:

- when an error is encountered very early during server startup.
- at the tail end of the `Drain` RPC, when the `Shutdown` request flag is true, to indicate the process can now exit.
- when a SQL liveness record is found to have been deleted or expired without heartbeat.

The way this works is that the internal component, given a reference to `*server.stopTrigger`, when it needs a server shutdown calls `signalStop()` and passes a "shutdown
request" (`serverctl.MakeShutdownRequest`).

At the very outer orchestration layer, a loop monitor a channel from the `stopTrigger` and processes shutdown requests. As of this writing, we have at least the following orchestration points:

1. the top-level shutdown logic in `cli/start.go`.
2. the top-level `testServer` shutdown monitor task (initiated from `(*server.testServer).Start()`)
3. the tenant server orchestration logic in `(*server.channelOrchestrator) startControlledServer()`.

Prior to this patch, a bug existed in `cli/start.go` as follows:

- the shutdown logic considered that any "shutdown reason" other than `ShutdownReasonServerStartupError` could be processed by performing a graceful drain on the server.
- meanwhile, in fact *none* of the current shutdown reasons can be correctly processed via a graceful drain (they all signal conditions that make the current server unhealthy and unable to shut down gracefully).
- as a result, if a shutdown request other than "server startup error" was received in `cli/start.go`, it would attempt a graceful drain and the graceful drain process would generally fail with all kinds of error messages in logs. It would also sometime hang, blocking the overall shutdown.

This patch fixes this by avoiding the graceful drain unless the shutdown request is for a graceful drain.

Through this, care is taken to ensure that if a hard shutdown is triggered after a graceful shutdown, the hard shutdown is also "delivered" to the orchestration. That allows hard shutdown to "catch up" with and shortcut graceful drains initiated earlier.

Release note (bug fix): A bug was fixed whereby `cockroach start` would sometimes incorrectly hang upon shutting down a server after encountering an internal error. This bug had been introduced some time in v22.x.

Fixes #108611.
Epic: CRDB-28893

### CODE REVIEW

**Review by stevendanna — APPROVED** [review]

**Comment by knz** [comment]
TFYR!

bors r=stevendanna

**Review by yuzefovich — COMMENTED** [review]
> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108612)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[`pkg/server/serverctl/shutdown.go` line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108612#-NbeeoH014XO7i757DWp:-NbeeoH014XO7i757DWq:b-xb6prf) ([raw file](https://github.com/cockroachdb/cockroach/blob/29668e190ee6550bc040ec72378f7eada0a40f01/pkg/server/serverctl/shutdown.go#L25)):*
> ```Go
> }
> 
> // Graceful determines whether the shutdown should be effected via a
> ```

nit: `s/Graceful/TerminateUsingGracefulDrain/`.
